File: src/main/java/com/ject/studytrip/auth/infra/repository/redis/RefreshTokenRedisRepositoryAdapter.java
package com.ject.studytrip.auth.infra.repository.redis;

import static com.ject.studytrip.global.common.constants.CacheKeyConstants.AUTH_REISSUE_TOKEN_PREFIX;

import com.ject.studytrip.auth.domain.repository.RefreshTokenRedisRepository;
import java.util.concurrent.TimeUnit;
import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Repository;

@Repository
@RequiredArgsConstructor
public class RefreshTokenRedisRepositoryAdapter implements RefreshTokenRedisRepository {
    private final RedisTemplate<String, String> redisTemplate;

    @Override
    public void saveRefreshToken(
            String memberId, String refreshToken, long refreshTokenExpireTime) {
        redisTemplate
                .opsForValue()
                .set(
                        AUTH_REISSUE_TOKEN_PREFIX.getValue() + refreshToken,
                        memberId,
                        refreshTokenExpireTime,
                        TimeUnit.MILLISECONDS);
    }

    @Override
    public boolean existsRefreshToken(String refreshToken) {
        return Boolean.TRUE.equals(
                redisTemplate.hasKey(AUTH_REISSUE_TOKEN_PREFIX.getValue() + refreshToken));
    }

    @Override
    public void deleteRefreshToken(String refreshToken) {
        redisTemplate.delete(AUTH_REISSUE_TOKEN_PREFIX.getValue() + refreshToken);
    }

    @Override
    public String findMemberIdByRefreshToken(String refreshToken) {
        return redisTemplate.opsForValue().get(AUTH_REISSUE_TOKEN_PREFIX.getValue() + refreshToken);
    }
}

File: src/main/kotlin/com/ject/studytrip/auth/infra/repository/redis/RefreshTokenRedisRepositoryAdapter.kt
package com.ject.studytrip.auth.infra.repository.redis

import com.ject.studytrip.auth.domain.repository.RefreshTokenRedisRepository
import com.ject.studytrip.global.common.constants.CacheKeyConstants.AUTH_REISSUE_TOKEN_PREFIX
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Repository
import java.util.concurrent.TimeUnit

@Repository
class RefreshTokenRedisRepositoryAdapter(
    private val redisTemplate: RedisTemplate<String, String>,
) : RefreshTokenRedisRepository {
    override fun saveRefreshToken(
        memberId: String,
        refreshToken: String,
        refreshTokenExpireTime: Long,
    ) {
        redisTemplate.opsForValue().set(buildKey(refreshToken), memberId, refreshTokenExpireTime, TimeUnit.MILLISECONDS)
    }

    override fun existsRefreshToken(refreshToken: String?): Boolean = redisTemplate.hasKey(buildKey(refreshToken)) == true

    override fun deleteRefreshToken(refreshToken: String?) {
        redisTemplate.delete(buildKey(refreshToken))
    }

    override fun findMemberIdByRefreshToken(refreshToken: String?): String? = redisTemplate.opsForValue().get(buildKey(refreshToken))

    private fun buildKey(refreshToken: String?): String = AUTH_REISSUE_TOKEN_PREFIX + refreshToken
}
